lib/charms/mongodb/v0/upgrade.py

```diff
@@ -65,30 +82,91 @@ def pre_upgrade_check(self) -> None:
             raise ClusterNotReadyError(message=default_message, cause="Cluster is not healthy")
 
         if not self.is_cluster_able_to_read_write():
-            raise ClusterNotReadyError(message=default_message, cause="Cluster cannot read/write")
+            raise ClusterNotReadyError(
+                message=default_message, cause="Cluster cannot read/write - please check logs"
+            )
 
         # Future PR - sharding based checks
 
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        if not self.is_cluster_able_to_read_write():
+            raise ClusterNotReadyError(
+                message="post-upgrade check failed and cannot safely upgrade",
+                cause="Cluster cannot read/write",
+            )
+
     @override
     def build_upgrade_stack(self) -> list[int]:
+        """Builds an upgrade stack, specifying the order of nodes to upgrade."""
+        if self.charm.is_role(Config.Role.CONFIG_SERVER):
+            # TODO implement in a future PR a stack for shards and config server
+            pass
+        elif self.charm.is_role(Config.Role.REPLICATION):
+            return self.get_replica_set_upgrade_stack()
+
+    def get_replica_set_upgrade_stack(self) -> list[int]:
```

Comment:
**question:** Is the reason this is a separate method because you will eventually have it split by `Config.Role`?
Like:
```python
get_replica_set_upgrade_stack()
get_config_server_upgrade_stack()
...
```

Reply:
yes exactly!